dev-docs/batch-system.rst

```diff
@@ -0,0 +1,784 @@
+cancelling batches with more than N failures if specified by the
+user. The service can be located on a preemptible machine, but we use
+a non-preemptible machine to minimize downtime, especially when the
+cluster is large. There can only be one driver service in existence at
+any one time. There is an Envoy side car container in the batch driver
+pod to handle TLS handshakes to avoid excess CPU usage of the batch
+driver.
+
+
+********************
+Instance Collections
+********************
+
+The batch driver maintains two different types of collections of
+workers. There are **pools** that are multi-tenant and have a
+dedicated worker type that is shared across all jobs. Pools can
+support both preemptible and nonpreemptible VMs. Right now, there are
+three types of machine types we support that correspond to low memory
+(~1GB memory / core), standard (~4GB memory / core), and high memory
+(~8GB memory / core) machines. These are correspondingly the
+"highcpu", "standard", and "highmem" pools. Each pool has its own
+scheduler and autoscaler. In addition, there's a single job private
+instance manager that creates a worker VM per job and is used if the
+worker requests a specific machine type. This is used commonly for
+jobs that require more memory than a 16 core machine can provide.
+
+
+**********
+Fair Share
+**********
+
+In order to avoid having one user starve other users from getting
+their jobs run, we use the following fair share algorithm. We start
+with the user who has the fewest cores running. We then allocate as
+many cores as possible that are live in the cluster until we reach the
+number of cores the next user has currently running. We then divide up
+the remaining cores equally amongst the two users until we reach the
+number of cores the next user has running. We repeat until we have
+either exhausted all free cores in the cluster or have satisfied all
+user resource requests. The query to get the number of ready cores in the fair
+share algorithm is fast because we aggregate across a global table
+``user_inst_coll_resources`` that has a limited number of rows
+maintaining counts of the number of ready cores per instance
+collection and user.
+
+
+**********
+Autoscaler
+**********
+
+At a high level, the autoscaler is in charge of figuring out how many
+worker VMs are required to run all of the jobs that are ready to run
+without wasting resources. The simplest autoscaler takes the number of
+ready cores total across all users and divides up that amount by the
+number of cores per worker to get the number of instances that are
+required. It then spins up a maximum of 10 instances each time the
+autoscaler runs to avoid cloud provider API rate limits. This approach
+works well for large workloads that have long running jobs. It is not
+very efficient if there's many short running jobs and the driver
+cannot handle the load from a large cluster or the workload is large
+but runs quickly.
```

Comment:
```suggestion
works well for large workloads that have long running jobs.

The autoscaler can produces more cores than the scheduler can keep
busy with work. This happens when there are many jobs with a short
execution time.
```
I get what you mean by efficiency but I think it's not obvious in this context that you're referring to minimizing the fraction of non-rented worker cores.